### PR TITLE
Don't execute a progression search query on projects without a FilePath.

### PR DESCRIPTION
**Pull request by brettfo**
This can happen in Venus projects which create temporary projects for each contained `.aspx` file.  E.g., if a Venus project contains `SomeVenusDocument.aspx`, then a hidden project with a name like `1_SomeNamespace_SomeVenusDocument.aspx` is created that has a null `FilePath` and only contains one document, `SomeVenusDocument.aspx.cs`.  This is a project system artifact to enable proper binding to the backing `.aspx.cs` file so that various IDE features (IntelliSense, completion, etc.) are enabled in the pseudo-HTML `.aspx` file.

This is in response to an internal bug where searching the Solution Explorer when Venus projects are loaded can cause VS to crash.  I was unable to repro this manually, but confirmed the cause of the crash by examining a private dump and confirming the existence of a project like I mentioned above: a specially-crafted name, null `FilePath`, and only one document pointing to the appropriate `.aspx.cs` file.

The crash was ultimately coming from [here](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Progression/GraphNodeIdCreation.cs,29) because the property `document.Project.FilePath` was null.

I didn't guard against `document.FilePath` being null because that scenario is explicitly not allowed and would already be caught by contract checks [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/DocumentInfo.cs,70).

The code path for C#/VB is shared, hence no VB-specific test.

Tagging @Pilchie @dpoeschl @jasonmalinowski @rchande @balajikris @basoundr @jmarolf @davkean for review.

### CODE REVIEW

File: src/EditorFeatures/Test/Workspaces/TestWorkspaceFactory_XmlConsumption.cs [hunk 1]
```diff
@@ -225,6 +225,11 @@ public static TestWorkspace CreateWorkspace(string xmlDefinition, bool completed
             if (projectElement.Attribute(FilePathAttributeName) != null)
             {
                 filePath = projectElement.Attribute(FilePathAttributeName).Value;
+                if (string.Compare(filePath, "null", StringComparison.Ordinal) == 0 || string.Compare(filePath, "Nothing", StringComparison.OrdinalIgnoreCase) == 0)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/Workspaces/TestWorkspaceFactory_XmlConsumption.cs]: I'd just pick one magic string to make it easier to grep for. Honestly, make it *really* magic -- have the property called TestWorkspaceFactory.NullPath, and then have the test do <%= TestWorkspaceFactory.NullPath %> in the XML literal.

**Comment by dpoeschl** [comment]
Does this mean that the .aspx.cs file isn't represented in progression? If so, is that what Dev12 did?

**Comment by jasonmalinowski** [comment]
:+1:. I'm surprised we can't get a manual repro on this though...did you try the usual adventures of web site vs. web app and razor vs. traditional?

File: src/EditorFeatures/Test/Workspaces/TestWorkspaceFactory_XmlConsumption.cs (hunk 1 above)
Comment (dpoeschl) [inline src/EditorFeatures/Test/Workspaces/TestWorkspaceFactory_XmlConsumption.cs]: Yes, I can imagine some unfortunate person in the future investigating some bug with files named "null" with no extension. :stuck_out_tongue: Just make the string a guid. :smile:

Comment (jasonmalinowski) [inline src/EditorFeatures/Test/Workspaces/TestWorkspaceFactory_XmlConsumption.cs]: Option B: allow empty string to mean null (unless we allow that distinguishing in the workspaces API...), or have a FilePathNull="true" attribute to make it really explicit.

**Comment by brettfo** [comment]
@dpoeschl: The .aspx.cs file is still represented in progression, but it's duplicated in the special project so you can get completion, etc. in the .aspx file.  This is to get around the fact that the .aspx section can do something like this: `<div runat="server" id="MyDiv">` but the backing generated file has the actual definition (e.g., `HtmlDiv MyDiv = new HtmlDiv();`) and that definition is needed to properly parse/bind during design time.

@jasonmalinowski: While I wasn't able to get a manual repro, the various bug reports on this all mention opening a project that was originally created in VS2013, but even creating projects there and opening them in VS2015 I couldn't get it to happen 'in the wild'.  On a related note, if anybody *does* get a manual repro for this, I'll buy them a drink (preferably one of the free ones from the kitchenette.)

**Comment by balajikris** [comment]
:+1: for the changes.

Note that typescript also does some interesting stuff - it is represented as 2 projects (one for the set of .ts files and another for the set of .cs files), though i'm not sure if any of them would have their filepath set to null. In your manual testing, try adding some .ts files to your asp.net test projects :smile:

**Comment by brettfo** [comment]
@balajikris: I added some .ts files as well and didn't see any oddities.

**Comment by dpoeschl** [comment]
:+1:

**Pull request merged**